AX: It's impossible to select a date in the date picker from <input type="date"> using the keyboard

twilco opened the pull request:
#### 110ae1a60592b7cdf1460fdce01c54aae21a8794
<pre>
AX: It&apos;s impossible to select a date in the date picker from &lt;input type=&quot;date&quot;&gt; using the keyboard
<a href="https://rdar.apple.com/161759633">rdar://161759633</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=299979">https://bugs.webkit.org/show_bug.cgi?id=299979</a>

Reviewed by Aditya Keerthi.

Prior to this commit, the NSDatePicker created by date inputs was entirely inaccessible to keyboard and assistive
technology users. The primary reason for this is that the NSDatePicker was not given first respondership, and thus
didn&apos;t receive any key events. This commit implements mechanisms that allowing managing keyboard focus state between
the date picker, and the backing date input. The interactity model for date inputs and their associated pickers are
as follows:

  - Pressing Space while focused on a date subfield (e.g. day, month, year field) moves focus into the picker.

  - Pressing Escape while focus is in the picker moves focus back to the previously focused subfield.

  - Clicking on a subfield with the mouse leaves focus on that subfield so you can increment / decrement using the
    arrow keys (matching shipping behavior and platform date pickers).

All of the above behavior is captured in a new layout test: picker-keyboard-accessibility.html

It&apos;s also important that the movement of keyboard focus into the picker does not cause any new web-exposed events to
be fired (e.g. blur). Shipping Safari, nor any other browser fire any blur or focus change events when opening the picker.
FocusController::dispatchEventsOnWindowAndFocusedElement is updated with this commit to avoid that, and new
test no-blur-when-opening-picker.html verifies we behave correctly on this front.

Finally, this commit also fixes an existing bug where pressing the space key to open the date picker scrolls the page
resulting in the picker appearing and immediately disappearing as a result of the scroll. New test space-does-not-scroll-page.html
verifies we behave correctly.

Tests: fast/forms/datetimelocal/datetimelocal-editable-components/no-blur-when-opening-picker.html
       fast/forms/datetimelocal/datetimelocal-editable-components/picker-keyboard-accessibility.html
       fast/forms/datetimelocal/datetimelocal-editable-components/space-does-not-scroll-page.html

* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/no-blur-when-opening-picker-expected.txt: Added.
* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/no-blur-when-opening-picker.html: Added.
* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/picker-keyboard-accessibility-expected.txt: Added.
* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/picker-keyboard-accessibility.html: Added.
* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/space-does-not-scroll-page-expected.txt: Added.
* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/space-does-not-scroll-page.html: Added.
* Source/WebCore/dom/Element.h:
(WebCore::Element::transferredFocusToPicker const):
(WebCore::Element::didSuppressBlurDueToPickerFocusTransfer):
* Source/WebCore/dom/FocusEvent.cpp:
(WebCore::FocusEvent::debugDescription const):
* Source/WebCore/dom/FocusEvent.h:
* Source/WebCore/html/BaseDateAndTimeInputType.cpp:
(WebCore::BaseDateAndTimeInputType::handleDOMActivateEvent):
(WebCore::BaseDateAndTimeInputType::didReceiveSpaceKeyFromControl):
(WebCore::BaseDateAndTimeInputType::setupDateTimeChooserParameters):
(WebCore::BaseDateAndTimeInputType::closeDateTimeChooser):
* Source/WebCore/html/BaseDateAndTimeInputType.h:
* Source/WebCore/html/InputType.cpp:
(WebCore::InputType::dispatchSimulatedClickIfActive):
(WebCore::InputType::dispatchSimulatedClickIfActive const): Deleted.
* Source/WebCore/html/InputType.h:
* Source/WebCore/html/shadow/DateTimeEditElement.cpp:
(WebCore::DateTimeEditElement::defaultEventHandler):
(WebCore::DateTimeEditElement::didFieldOwnerTransferFocusToPicker):
(WebCore::DateTimeEditElement::didSuppressBlurDueToPickerFocusTransfer):
* Source/WebCore/html/shadow/DateTimeEditElement.h:
* Source/WebCore/html/shadow/DateTimeFieldElement.cpp:
(WebCore::DateTimeFieldElement::transferredFocusToPicker const):
(WebCore::DateTimeFieldElement::didSuppressBlurDueToPickerFocusTransfer):
* Source/WebCore/html/shadow/DateTimeFieldElement.h:
* Source/WebCore/page/FocusController.cpp:
(WebCore::dispatchEventsOnWindowAndFocusedElement):
* Source/WebCore/platform/DateTimeChooserParameters.h:
* Source/WebKit/Platform/spi/mac/AppKitSPI.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/mac/WebDateTimePickerMac.h:
* Source/WebKit/UIProcess/mac/WebDateTimePickerMac.mm:
(-[WKDateTimePicker initWithParams:inView:]):
(-[WKDateTimePicker showPicker:]):
(-[WKDateTimePicker updatePicker:]):
(-[WKDateTimePicker invalidate]):
(-[WKDateTimePicker handleEscapeKey]):
(-[WKDateTimePicker didChooseDate:]):
(-[WKDateTimePicker wasActivatedByKeyboard]):
(-[WKEscapeHandlingDatePicker setDateTimePicker:]):
(-[WKEscapeHandlingDatePicker dateTimePicker]):
(-[WKEscapeHandlingDatePicker keyDown:]):
(-[WKEscapeHandlingDatePicker acceptsFirstResponder]):

Canonical link: <a href="https://commits.webkit.org/301432@main">https://commits.webkit.org/301432@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/062ccd3afa1c496e4de348c944e8f60009c49ced

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125950 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/45610 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/36380 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/132816 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/77806 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/774a201f-257c-44c5-9f09-272ec57504ee) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/127821 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/46293 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54167 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95955 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/64050 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/a600b3ef-aedd-41b0-bc1c-c74762d2a7f5) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/128898 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/37026 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/112639 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76444 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/14721bdf-9c33-4f67-9b47-9bcb73362b30) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35934 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/30824 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/76286 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106808 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/31037 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/135502 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/52732 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/40462 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104430 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53181 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108850 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104150 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26525 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/49525 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27853 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/50110 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/52626 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/58441 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51970 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55318 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/53667 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/111534456e5f8cef41889aa4da9510183ceb180d. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51648

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/703e4cec20353e311c3a9d48b077eb45f090a0c7. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51648

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/f6cadc6772bd237025d755f8aa4ee955ef422d4e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51648

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/794322a1927a8c188d0c3a8629e9d04b4d650647. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51648

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/1a4c1735e44c50c94228993f6db3345087696fbf. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51648

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/3e50bce33f6acdcb668ae64d768581f7e72f620b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51648

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/3ddfbff64428e04ebef86376fac013dc423f22aa. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51648

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/3fc2b9ad32133d54a27aaf302d858bb31e5e32f7. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51648

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/3895c2b8f8675ae54eb4ae7f7b9c89f3fbfe1df4. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51648

twilco commented on Source/WebCore/html/BaseDateAndTimeInputType.cpp:
```diff
@@ -325,12 +325,38 @@ void BaseDateAndTimeInputType::setValue(const String& value, bool valueChanged,
+    }
+
+    m_pickerWasActivatedByKeyboard = true;
+    // Focus is going to move from a subfield in the input to the chooser, so
+    // preserve the chooser in the blur that said focus movement will cause.
+    m_preserveChooserInNextBlur = true;
```
Good catch! I tested other browsers, and no-one else emits a web-observable `blur` when using the keyboard to open the `<input type="date">` picker. I think it would be reasonably easy to avoid emitting a blur when moving focus into the picker, so I will try that first thing tomorrow, and add an additional layout test proving we don't emit an extra blur.

twilco committed:
AX: It's impossible to select a date in the date picker from <input type="date"> using the keyboard
rdar://161759633
https://bugs.webkit.org/show_bug.cgi?id=299979

Reviewed by Aditya Keerthi.

Prior to this commit, the NSDatePicker created by date inputs was entirely inaccessible to keyboard and assistive
technology users. The primary reason for this is that the NSDatePicker was not given first respondership, and thus
didn't receive any key events. This commit implements mechanisms that allowing managing keyboard focus state between
the date picker, and the backing date input. The interactity model for date inputs and their associated pickers are
as follows:

  - Pressing Space while focused on a date subfield (e.g. day, month, year field) moves focus into the picker.

  - Pressing Escape while focus is in the picker moves focus back to the previously focused subfield.

  - Clicking on a subfield with the mouse leaves focus on that subfield so you can increment / decrement using the
    arrow keys (matching shipping behavior and platform date pickers).

All of the above behavior is captured in a new layout test: picker-keyboard-accessibility.html

It's also important that the movement of keyboard focus into the picker does not cause any new web-exposed events to
be fired (e.g. blur). Shipping Safari, nor any other browser fire any blur or focus change events when opening the picker.
FocusController::dispatchEventsOnWindowAndFocusedElement is updated with this commit to avoid that, and new
test no-blur-when-opening-picker.html verifies we behave correctly on this front.

Finally, this commit also fixes an existing bug where pressing the space key to open the date picker scrolls the page
resulting in the picker appearing and immediately disappearing as a result of the scroll. New test space-does-not-scroll-page.html
verifies we behave correctly.

Tests: fast/forms/datetimelocal/datetimelocal-editable-components/no-blur-when-opening-picker.html
       fast/forms/datetimelocal/datetimelocal-editable-components/picker-keyboard-accessibility.html
       fast/forms/datetimelocal/datetimelocal-editable-components/space-does-not-scroll-page.html

* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/no-blur-when-opening-picker-expected.txt: Added.
* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/no-blur-when-opening-picker.html: Added.
* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/picker-keyboard-accessibility-expected.txt: Added.
* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/picker-keyboard-accessibility.html: Added.
* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/space-does-not-scroll-page-expected.txt: Added.
* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/space-does-not-scroll-page.html: Added.
* Source/WebCore/dom/Element.h:
(WebCore::Element::transferredFocusToPicker const):
(WebCore::Element::didSuppressBlurDueToPickerFocusTransfer):
* Source/WebCore/dom/FocusEvent.cpp:
(WebCore::FocusEvent::debugDescription const):
* Source/WebCore/dom/FocusEvent.h:
* Source/WebCore/html/BaseDateAndTimeInputType.cpp:
(WebCore::BaseDateAndTimeInputType::handleDOMActivateEvent):
(WebCore::BaseDateAndTimeInputType::didReceiveSpaceKeyFromControl):
(WebCore::BaseDateAndTimeInputType::setupDateTimeChooserParameters):
(WebCore::BaseDateAndTimeInputType::closeDateTimeChooser):
* Source/WebCore/html/BaseDateAndTimeInputType.h:
* Source/WebCore/html/InputType.cpp:
(WebCore::InputType::dispatchSimulatedClickIfActive):
(WebCore::InputType::dispatchSimulatedClickIfActive const): Deleted.
* Source/WebCore/html/InputType.h:
* Source/WebCore/html/shadow/DateTimeEditElement.cpp:
(WebCore::DateTimeEditElement::defaultEventHandler):
(WebCore::DateTimeEditElement::didFieldOwnerTransferFocusToPicker):
(WebCore::DateTimeEditElement::didSuppressBlurDueToPickerFocusTransfer):
* Source/WebCore/html/shadow/DateTimeEditElement.h:
* Source/WebCore/html/shadow/DateTimeFieldElement.cpp:
(WebCore::DateTimeFieldElement::transferredFocusToPicker const):
(WebCore::DateTimeFieldElement::didSuppressBlurDueToPickerFocusTransfer):
* Source/WebCore/html/shadow/DateTimeFieldElement.h:
* Source/WebCore/page/FocusController.cpp:
(WebCore::dispatchEventsOnWindowAndFocusedElement):
* Source/WebCore/platform/DateTimeChooserParameters.h:
* Source/WebKit/Platform/spi/mac/AppKitSPI.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/mac/WebDateTimePickerMac.h:
* Source/WebKit/UIProcess/mac/WebDateTimePickerMac.mm:
(-[WKDateTimePicker initWithParams:inView:]):
(-[WKDateTimePicker showPicker:]):
(-[WKDateTimePicker updatePicker:]):
(-[WKDateTimePicker invalidate]):
(-[WKDateTimePicker handleEscapeKey]):
(-[WKDateTimePicker didChooseDate:]):
(-[WKDateTimePicker wasActivatedByKeyboard]):
(-[WKEscapeHandlingDatePicker setDateTimePicker:]):
(-[WKEscapeHandlingDatePicker dateTimePicker]):
(-[WKEscapeHandlingDatePicker keyDown:]):
(-[WKEscapeHandlingDatePicker acceptsFirstResponder]):

Canonical link: https://commits.webkit.org/301432@main
